docs/design-documents/features/storage/SecureStore/SecureStore_design.md

```diff
@@ -0,0 +1,448 @@
+	- If authentication flag set, calculate CMAC on `dest_buf`, using `_auth_handle` handle
+	- If encrypt flag set, decrypt `dest_buf` (in place)  using `_enc_handle` handle
+	- Decrement `data_size` by `chunk_size`
+- Call `_underlying_kv` `get` API with on a local `read_cmac` variable, size 16
+- Generate CMAC on local `cmac` variable 
+- If `read_cmac` doesn't match `cmac`, return "data corrupt error"
+- If rollback protection flag set
+	- Call `_rbp_kv` `get` API on a local `rbp_cmac` variable, key is `key`, size 16
+	- If `rbp_cmac` doesn't match `cmac`, clear `buffer` and return "RBP authentication" error
+- Deinitialize and free `auth_handle` and `enc_handle`
+- Release `_mutex`
+- Return OK   
+
+**get_info function**
+
+Header:  
+`virtual int get_info(const char *key, info_t *info);`
+
+Pseudo code:  
+- if not `_is_initialized` return error
+- Call `get` API with `key` and 0 in `buffer_size` parameter
+- If failed, return error code  
+- Call `_underlying_kv` `get` API with `metadata` size and `key`
+- Fill fields in `info` according to `metadata`
+- Return OK   
+
+**remove function**
+
+Header:  
+`virtual int remove(const char *key);`
+
+Pseudo code:  
+- if not `_is_initialized` return error
+- Take `_mutex` 
+- Call `_underlying_kv` `get` API with `metadata` size and `key`
+- If not found, return "Not found" error
+- If write once flag set, return "Already exists" error
+- Call `_underlying_kv` `remove` API with `key`
+- If rollback protect flag set, Call `_rbp_kv` `remove` API with `key` as key
+- Return OK
+
+### Incremental set APIs
+
+**set_start function**
+
+Header:  
+`virtual int set_start(set_handle_t *handle, const char *key, size_t final_data_size, uint32_t create_flags);`
+
+Pseudo code:
+- Take `_mutex`  
+- Allocate an `inc_set_handle_t` and assign in handle
+- if flags include write once flag
+	- Call `_underlying_kv` `get_info` API
+	- If key exists, return "already exists" error
+	- Call `_rbp_kv` `get` API with `key` as key. If key exists, return "already exists" error
+- If encrypt flag set
+	- Derive a key from device key and `key` as salt (trimmed to 32 bytes with "ENC" as prefix)
+	- Using TLS entropy function on `_entropy` handle, randomly generate `nonce` field
+	- Allocate and initialize `enc_handle` AES-CTR handle field with derived key and `nonce` field 
+- Fill all available fields in `metadata` 
+- If authentication flag set
+	- Derive a key from device key and `key` as salt (trimmed to 32 bytes with "AUTH" as prefix)
+	- Allocate and initialize `auth_handle` CMAC calculation handle field with derived key 
+	- Using `auth_handle` handle, calculate CMAC on `key` & `metadata`
+- Call `_underlying_kv` `set_start` API 	    
+- Call `_underlying_kv` `set_add_data` API with `metadata` field 	    
+- Return OK   
+
+**set_add_data function**
+
+Header:  
+`virtual int set_add_data(set_handle_t handle, const void *value_data, size_t data_size);`
```

Comment:
Should add documentation on the `set_add_data()` function letting the user know that the data is not protected until the `set_finalize()` function is called

Reply:
It is important to state that the stored record is not valid until `set_finalized` is called. `set_add_data` only adds a partial data chunk to the data. One cannot attempts to read the data before `set_finalize` is called, as the call is protected by a mutex. If we get a power failure between these calls, reading the data would fail on authentication, as CMAC is written last. If encryption is required, then `set_add_data` will write the data encrypted, so there's no fear here that the written data is exposed.

Reply:
correct. can we add that to the documentation of the functions?

Reply:
Sure. Will add.